pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,416 @@
+        self.lib = AscentWrappers()
+        self.ascent_ptr = self.lib.ascent_create(None)
+
+        self.ascent_config = ConduitNode(self.conduit)
+        self.ascent_config['mpi_comm'] = comm.py2f()
+        self.ascent_config['runtime/type'] = 'ascent'
+        vtkm_backend = self.cfg.get(cfgsect, 'vtkm-backend', 'serial')
+        self.ascent_config['runtine/vtkm/backend'] = vtkm_backend
+
+        # Open ascent
+        self.lib.ascent_open(self.ascent_ptr, self.ascent_config)
+
+        # Pre configure scenes and pipelines
+        self.actions = ConduitNode(self.conduit)
+        self._add_scene = self.actions.append()
+        self._add_scene['action'] = 'add_scenes'
+        self._add_scene['scenes'] = self.scenes
+
+        self._add_pipeline = self.actions.append()
+        self._add_pipeline['action'] = 'add_pipelines'
+        self._add_pipeline['pipelines'] = self.pipelines
+
+    def __del__(self):
+        if getattr(self, 'ascent_ptr', None):
+            self.lib.ascent_close(self.ascent_ptr)
+
+    def _config_field(self, cfgsect):
+        cons = self.cfg.items_as('constants', float)
+
+        for k in self.cfg.items(cfgsect, prefix='field-'):
+            field = k.removeprefix('field-')
+            configs = self.cfg.getliteral(cfgsect, k)
+
+            for kc, vc in configs.items():
+                if kc.startswith('field_expr'):
+                    self._add_field_expr(field, kc, vc, cons)
+
+    def _config_scene(self, cfgsect):
+        cons = self.cfg.items_as('constants', float)
+        self._scene_cfg = scene_cfg = {}
+        self.scenes = ConduitNode(self.conduit)
+
+        for k in self.cfg.items(cfgsect, prefix='scene-'):
+            sn = k.removeprefix('scene-')
+            scene_cfg[sn] = cfg = self.cfg.getliteral(cfgsect, k)
+            field = cfg['field']
+
+            self._fields_read.add(field)
+
+            for kc, vc in cfg.items():
+                if kc == 'image_name':
+                    pass
+                elif kc.startswith('field_expr'):
+                    self._add_field_expr(field, kc, vc, cons)
+                elif kc == 'pipeline':
+                    self.scenes[f's_{sn}/plots/p1/pipeline'] = f'pl_{vc}'
+                elif kc.startswith('render-'):
+                    ropt = kc.removeprefix('render-')
+                    self.scenes[f's_{sn}/renders/r1/{ropt}'] = vc
+                else:
+                    self.scenes[f's_{sn}/plots/p1/{kc}'] = vc
+
+    def _config_pipeline(self, cfgsect):
+        self._pipel_cfg = pipel_cfg = {}
+        self.pipelines = pl = ConduitNode(self.conduit)
+
+        for k in self.cfg.items(cfgsect, prefix='pipeline-'):
+            pn = k.removeprefix('pipeline-')
+            cfg = self.cfg.getliteral(cfgsect, k)
+            pipel_cfg[pn] = cfg = [cfg] if isinstance(cfg, dict) else cfg
+
+            for j, filter in enumerate(cfg):
+                params = ConduitNode(self.conduit)
+                ptype = None
```

Comment:
`pl[f'pl_{pn}/f{j}/type'] = ptype = filter.pop('type')`?

Get the exception if there is no type and avoids the special case in the loop below.